lib/PrometheusPlugin.js

```diff
@@ -0,0 +1,244 @@
+
+/**
+ * @class PrometheusPlugin
+ *
+ * @property {KuzzlePluginContext}     context        Kuzzle plugin context
+ * @property {Object.<string, string>} hooks          Kuzzle plugin hooks
+ * @property {Routes}                  routes         Kuzzle plugin routes
+ * @property {Controllers}             controllers    Kuzzle plugin controllers
+ * @property {Prometheus.Registry}     registry       Prometheus Registry
+ * @property {Object.<string, Object>} kuzzleMetrics  Prometheus metrics dedicated for Kuzzle
+ * @property {String}                  key            Key used when pushing Prometheus Registry to Redis
+ * @property {String}                  hostname       Hostname
+ * @property {String}                  MAC            Host MAC address
+ * @property {String}                  IP             Host IPv4 adress
+ *
+ * @externs
+ */
+class PrometheusPlugin {
+  /**
+   * PrometheusPlugin constructor
+   *
+   * @returns {PrometheusPlugin}
+   */
+  constructor() {
+    this.controllers = {
+      prometheus: {
+        metrics: 'metrics'
+      }
+    };
+
+    this.routes = [
+      {
+        verb: 'get',
+        url: '/metrics',
+        controller: 'prometheus',
+        action: 'metrics'
+      }
+    ];
+  }
+
+  /**
+   * Plugin initialisation
+   *
+   * @param {Object}              config   Plugin configuration from kuzzlerc file
+   * @param {KuzzlePluginContext} context
+   */
+  async init(config, context) {
+    this.config = { ...config };
+    this.context = context;
+
+    /**
+     * Using MAC address to generate Kuzzle Cluster node
+     * specific Redis key. This can be spoofed but
+     * if this the case, falsified Prometheus metrics are
+     * not your main problem.
+     */
+    this.hostname = os.hostname();
+    this.MAC = await promisify(require('macaddress').one)();
+    this.IP = getIPFromMAC(this.MAC);
+    this.key = `kuzzle_prometheus_${this.hostname}_${this.nodeMAC}_${this.IP}`;
+
+    this.registry = Prometheus.Registry.globalRegistry;
+    this.registry.setDefaultLabels({
+      nodeHost: this.hostname,
+      nodeMAC: this.MAC,
+      nodeIP: this.IP
+    });
+
+    this.kuzzleMetrics = {
+      requests: new Prometheus.Histogram({
+        name: `kuzzle_requests`,
+        help: `Kuzzle monitoring succeeded requests`,
+        buckets: Prometheus.exponentialBuckets(0.0001, 1.5, 36),
+        labelNames: ['controller', 'action', 'event', 'status', 'protocol']
+      }),
+      rooms: new Prometheus.Gauge({
+        name: `kuzzle_active_rooms`,
+        help: `Kuzzle active rooms counter`
+      })
+    };
+
+    this.hooks = {
+      'room:new': this.recordRooms.bind(this),
+      'room:remove': this.recordRooms.bind(this),
+      'request:onSuccess': this.recordRequests.bind(this),
+      'request:onError': this.recordRequests.bind(this)
+    };
+
+    this.systemMetricsJob = Prometheus.collectDefaultMetrics({
+      register: this.register,
+      timeout: this.config.systemMetricsInterval || 5000
```

Comment:
It's more readable to define your default configuration value in the constructor and the merge it with the provided config